### PR TITLE
Added dtw functions

### PR DESCRIPTION
**Pull request by luisa-vieira**

### CODE REVIEW

File: src/LineageTree/lineageTree.py [hunk 1]
```diff
@@ -25,10 +25,13 @@
 import matplotlib.pyplot as plt
 import networkx as nx
 import numpy as np
+from scipy.interpolate import InterpolatedUnivariateSpline
 
 # from edist.uted import uted
+
 from scipy.spatial import Delaunay
 from scipy.spatial import cKDTree as KDTree
+from sklearn.metrics.pairwise import euclidean_distances
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:34]: Is there another function to compute the distance matrix? As of now, `sklearn` is not necessary to make lineage tree work and I'd like to keep it that way if possible.

If there is an equivalent function in `numpy` or `scipy` please use it.

File: src/LineageTree/lineageTree.py [hunk 2]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
             return list(r)
         return final_nodes
 
+    @staticmethod
+    def __dp(dist_mat, start_d=0, back_d=0):
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2407]:
```python
def __dp(dist_mat: np.ndarry, start_d: int=0, back_d: int=0) -> (((int, int), ...), np.ndarry):
```
Or something like that

File: src/LineageTree/lineageTree.py [hunk 3]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+                ]  # deletion (2)
+                i_penalty = np.argmin(penalty)
+                cost_mat[i + 1, j + 1] = dist_mat[i, j] + penalty[i_penalty]
+                traceback_mat[i, j] = i_penalty
+
+        min_index1 = np.argmin(cost_mat[N - back_d :, M])
+        min_index2 = np.argmin(cost_mat[N, M - back_d :])
+
+        if (
+            cost_mat[N - back_d + min_index1, M]
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2452]: I rather always use the symbole `<` so the smaller element is always to the left.

File: src/LineageTree/lineageTree.py [hunk 4]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+            > cost_mat[N, M - back_d + min_index2]
+        ):
+            i = N - 1
+            j = M - back_d + min_index2 - 1
+        else:
+            i = N - back_d + min_index1 - 1
+            j = M - 1
+
+        path = [(i, j)]
+        while (i > start_d and j > 0) or (i > 0 and j > start_d):
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2462]: Same here, please invert the comparison sign

File: src/LineageTree/lineageTree.py [hunk 5]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+        # Fill the cost matrix while keeping traceback information
+        traceback_mat = np.zeros((N, M))
+
+        cost_mat[: start_d + 1, 0] = 0
+        cost_mat[0, : start_d + 1] = 0
+
+        cost_mat[N - back_d :, M] = 0
+        cost_mat[N, M - back_d :] = 0
+
+        square_root = np.sqrt((slope**2) + ((-1) ** 2))
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2535]: maybe don't compute `(-1)**2`? Just directly put `1`

File: src/LineageTree/lineageTree.py [hunk 6]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+                # Insertion
+                i -= 1
+            elif tb_type == 2:
+                # Deletion
+                j -= 1
+            path.append((i, j))
+
+        # Strip infinity edges from cost_mat before returning
+        cost_mat = cost_mat[1:, 1:]
+        return (path[::-1], cost_mat)
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2665]: There is a lot of repeated code between the three version of `dp`, maybe you could refactor?

File: src/LineageTree/lineageTree.py [hunk 7]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+
+    @staticmethod
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2667]: Put a reference to where we found it.

File: src/LineageTree/lineageTree.py [hunk 8]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+        t = -R @ centroid_A + centroid_B
+
+        return R, t
+
+    def __interpolate(self, track1, track2, threshold):
+        """
+        Interpolate two series that have different lengths
+
+            Args:
+                track1 (list) and track2 (list): list of nodes in cell cycle
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2716]:
```python
"""
track1 (list): list of nodes of the first cell cycle to compare
track2 (list): list of nodes of the second cell cycle to compare
"""
```

File: src/LineageTree/lineageTree.py [hunk 9]
```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+            len(track1) <= threshold or len(track2) <= threshold
+        ):
+            return track1_pos, track2_pos
+        # Both tracks have the same length but one or more sizes are above the threshold
+        elif (
+            len(track1) == len(track2)
+            and (len(track1) > threshold or len(track2) > threshold)
+            or len(track1) != len(track2)
+            and (len(track1) > threshold or len(track2) > threshold)
+        ):
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2740]: You are doing weird comparisons:
```python
            len(track1) == len(track2)
            and (len(track1) > threshold or len(track2) > threshold)
            or len(track1) != len(track2)
            and (len(track1) > threshold or len(track2) > threshold)
```
is equivalent to:
```python
            (len(track1) > threshold or len(track2) > threshold)
            and (
                        len(track1) != len(track2)
                        or len(track1) == len(track2)
            )
```
and because 
```python
            (
                        len(track1) != len(track2)
                        or len(track1) == len(track2)
            )
```
is always true then it is only:
```python
(len(track1) > threshold or len(track2) > threshold)
```
am I correct and if yes is it normal?

**Review by leoguignard — CHANGES_REQUESTED** [review]
1) Please do a git pull and double check that everything is working, I've made minor changes
2) Please add the type of the parameters and the outputs for all function

File: src/LineageTree/lineageTree.py (hunk 1 above)
Comment (luisa-vieira) [reply src/LineageTree/lineageTree.py:34]: yes, we can use eucledian from scipy

File: src/LineageTree/lineageTree.py [hunk 10]
```diff
@@ -2400,6 +2403,617 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+        x1 = max(0, i - j) / 2
+        x2 = (i + min(i, j)) / 2
+        y1 = max(0, j - i) / 2
+        y2 = (j + min(i, j)) / 2
+        slope = (y1 - y2) / (x1 - x2)
+        intercept = y1 - slope * x1
+        return slope, intercept
+    
+    @staticmethod
+    # Reference: https://github.com/kamperh/lecture_dtw_notebook/blob/main/dtw.ipynb
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2428]: I think it's better to put the comment before the `@staticmethod`

File: src/LineageTree/lineageTree.py [hunk 11]
```diff
@@ -2400,6 +2403,617 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+            if condition:
+                penalty = [
+                    cost_mat[i, j],  # match (0)
+                    cost_mat[i, j + 1],  # insertion (1)
+                    cost_mat[i + 1, j], # deletion (2)
+                ]
+                i_penalty = np.argmin(penalty)
+                cost_mat[i + 1, j + 1] = dist_mat[i, j] + penalty[i_penalty]
+                traceback_mat[i, j] = i_penalty
+        # for i in range(N):
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2477]: If you don't need it anymore, please remove the commented lines of code

File: src/LineageTree/lineageTree.py [hunk 12]
```diff
@@ -2400,6 +2403,617 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+    #             i -= 1
+    #         elif tb_type == 2:
+    #             # Deletion
+    #             j -= 1
+    #         path.append((i, j))
+
+    #     # Strip infinity edges from cost_mat before returning
+    #     cost_mat = cost_mat[1:, 1:]
+    #     return (path[::-1], cost_mat)
+
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2683]: Please remove the unnecessary commented lines of code

File: src/LineageTree/lineageTree.py [hunk 13]
```diff
@@ -2400,6 +2403,617 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+            len(track1) <= threshold or len(track2) <= threshold
+        ):
+            return track1_pos, track2_pos
+        # Both tracks have the same length but one or more sizes are above the threshold
+        elif (
+            len(track1) == len(track2)
+            and (len(track1) > threshold or len(track2) > threshold)
+            or len(track1) != len(track2)
+            and (len(track1) > threshold or len(track2) > threshold)
+        ):
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2759]: You didn't change that?

**Review by leoguignard — CHANGES_REQUESTED** [review]

File: src/LineageTree/lineageTree.py [hunk 14]
```diff
@@ -2485,28 +2485,32 @@ def __dp(dist_mat: np.ndarray, start_d: int=0, back_d: int=0, w: int=0, slope=No
         ):
             i = N - 1
             j = M - back_d + min_index2 - 1
+            final_cost = cost_mat[N, M - back_d + min_index2]
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2488]: Why do you have `i` and `j` declared and not used?

File: src/LineageTree/lineageTree.py [hunk 15]
```diff
@@ -2662,13 +2666,13 @@ def calculate_dtw(
         if fast:
             if centered_band:
                 slope, intercept = self.__calculate_diag_line(dist_mat)
-                path, cost_mat = self.__dp(dist_mat, start_d, back_d, w, slope, intercept, use_absolute=False)
+                path, cost_mat, final_cost = self.__dp(dist_mat, start_d, back_d, w, slope, intercept, use_absolute=False)
             else:
-                path, cost_mat = self.__dp(dist_mat, start_d, back_d, w, use_absolute=True)
+                path, cost_mat, final_cost = self.__dp(dist_mat, start_d, back_d, w, use_absolute=True)
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2671]: Is it normal that the `use_absolute` variable is hard coded?

File: src/LineageTree/lineageTree.py [hunk 16]
```diff
@@ -2444,9 +2448,7 @@ def __dp(dist_mat: np.ndarry, start_d: int=0, back_d: int=0, w: int=0, slope=Non
         w_limit = max(w, abs(N - M))  # Calculate the Sakoe-Chiba band width
 
         # Initialize the cost matrix
-        cost_mat = np.zeros((N + 1, M + 1))
-        cost_mat[0, 1:] = np.inf
-        cost_mat[1:, 0] = np.inf
+        cost_mat = np.full((N + 1, M + 1), np.inf)
```
Comment (leoguignard) [inline src/LineageTree/lineageTree.py:2451]: That's indeed much better

**Review by leoguignard — CHANGES_REQUESTED** [review]
Have you ran black on the latest version of the code?

File: src/LineageTree/lineageTree.py (hunk 14 above)
Comment (luisa-vieira) [reply src/LineageTree/lineageTree.py:2488]: i and j are used below to calculate the path
the final cost is the point where the path ends, if there's no delay, it's (N-1, M-1), but it can change when we add delay, so I created the variable inside the if

File: src/LineageTree/lineageTree.py (hunk 15 above)
Comment (luisa-vieira) [reply src/LineageTree/lineageTree.py:2671]: what do you mean, hard coded?

File: src/LineageTree/lineageTree.py [hunk 17]
```diff
@@ -2400,6 +2403,445 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+                "-",
+                label=f"root = {nodes1}",
+            )
+            ax.plot(
+                pos_cycle2[:, 0],
+                pos_cycle2[:, 1],
+                "-",
+                label=f"root = {nodes2}",
+            )
+
```
Comment (luisa-vieira) [inline src/LineageTree/lineageTree.py:2818]: I don't know how to best import this library

File: src/LineageTree/lineageTree.py [hunk 18]
```diff
@@ -2400,6 +2403,445 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+            cost_mat_p=True,
+        )
+
+        fig = plt.figure(figsize=(10, 6))
+        if three_dim:
+            ax = fig.add_subplot(1, 1, 1, projection="3d")
+        else:
+            ax = fig.add_subplot(1, 1, 1)
+
+        if three_dim:
```
Comment (luisa-vieira) [inline src/LineageTree/lineageTree.py:2790]: Is it ok to add a function inside a function? Otherwise I'd have to type this three times

Comment (leoguignard) [reply src/LineageTree/lineageTree.py:2790]: Why do you have to write it inside? You could write it in the class rather no?

File: src/LineageTree/lineageTree.py (hunk 17 above)
Comment (leoguignard) [reply src/LineageTree/lineageTree.py:2818]: which library?

File: src/LineageTree/lineageTree.py [hunk 19]
```diff
@@ -2400,6 +2403,445 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+        threshold: int=1000,
+        regist: bool=True,
+        start_d: int=0,
+        back_d: int=0,
+        fast: bool=False,
+        w: int=0,
+        centered_band: bool=True,
+        three_dim: bool=True,
+        alig: bool=False,
+    ) -> (float, plt.figure):
```
Comment (luisa-vieira) [inline src/LineageTree/lineageTree.py:2750]: What if we deleted the three_dim variable and only kept the projection where the user could write:
'3d' : for the 3d visualization
'xy' or None (default) : 2D projection of axis x and y
'xz' : 2D projection of axis x and z
'yz' : 2D projection of axis y and z
'pca' : PCA projection
Because the way that it's now three_dim default is True, so if the user only types plot_dtw_trajectory(nodes1, nodes2, projection = 'xy') he's still gonna get the 3D plot, so I think it can be confusing.

File: src/LineageTree/lineageTree.py (hunk 17 above)
Comment (luisa-vieira) [reply src/LineageTree/lineageTree.py:2818]: from sklearn.decomposition import PCA

File: src/LineageTree/lineageTree.py [hunk 20]
```diff
@@ -2400,6 +2403,445 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
...
+                else:
+                    ax.plot(x_pos, y_pos, "k--", color="grey")
+
+        ax.set_aspect("equal")
+        ax.set_xlabel("x position")
+        ax.legend()
+        ax.set_ylabel("y position")
+        fig.tight_layout()
+
+        return distance, fig
```
Comment (luisa-vieira) [inline src/LineageTree/lineageTree.py:2840]: the alignment doesn't work for the pca projection, so I added this warning

File: src/LineageTree/lineageTree.py (hunk 19 above)
Comment (leoguignard) [reply src/LineageTree/lineageTree.py:2792]: I think you're right. Do it that way then

File: src/LineageTree/lineageTree.py (hunk 20 above)
Comment (leoguignard) [reply src/LineageTree/lineageTree.py:2962]: Fair enough. You should use the warning then instead of print

File: src/LineageTree/lineageTree.py (hunk 17 above)
Comment (leoguignard) [reply src/LineageTree/lineageTree.py:2939]: Because it's a large library and it's only needed there, you can import it within the function.

You should then put a warning message if they want to use pca but don't have sklearn.

To check if they have it you can do
```python
try:
    from sklearn.decomposition import PCA
except ImportError:
    Warning("scikit-learn is not installed, the PCA orientation cannot be used. You can install scikit-learn with pip install")
```
Or something like that.

**Review by leoguignard — APPROVED** [review]
